core/src/main/java/io/grpc/internal/InternalHandlerRegistry.java

```diff
@@ -83,7 +86,9 @@ InternalHandlerRegistry build() {
           map.put(method.getMethodDescriptor().getFullMethodName(), method);
         }
       }
-      return new InternalHandlerRegistry(services.values(), Collections.unmodifiableMap(map));
+      return new InternalHandlerRegistry(
+          Collections.unmodifiableList(new ArrayList<ServerServiceDefinition>(services.values())),
```

Comment:
you might consider making this ordered, since it makes unit tests that verify the results easier.